File: M2 Atividade/IIR_Rejeita_Faixa/rejeita faixa.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import freqz
import os


def calculate_notch_coefficients(fs, f0, q):
    """
    Calcula os coeficientes de um filtro IIR rejeita-faixa.

    Parâmetros:
    - fs: Taxa de amostragem (Hz)
    - f0: Frequência central do notch (Hz)
    - q: Fator de qualidade

    Retorna:
    - Coeficientes (b, a): Numerador e denominador do filtro
    """
    omega = 2 * np.pi * f0 / fs  # Frequência angular
    alpha = np.sin(omega) / (2 * q)

    # Coeficientes do numerador
    b0 = 1
    b1 = -2 * np.cos(omega)
    b2 = 1

    # Coeficientes do denominador
    a0 = 1 + alpha
    a1 = -2 * np.cos(omega)
    a2 = 1 - alpha

    # Normalizar coeficientes
    b = [b0 / a0, b1 / a0, b2 / a0]
    a = [1, a1 / a0, a2 / a0]

    return b, a


def apply_filter(input_signal, b, a):
    """
    Aplica um filtro IIR a um sinal de entrada.

    Parâmetros:
    - input_signal: Sinal de entrada (array)
    - b: Coeficientes do numerador
    - a: Coeficientes do denominador

    Retorna:
    - Sinal filtrado
    """
    y = np.zeros_like(input_signal, dtype=float)
    for n in range(2, len(input_signal)):
        y[n] = (
            b[0] * input_signal[n]
            + b[1] * input_signal[n - 1]
            + b[2] * input_signal[n - 2]
            - a[1] * y[n - 1]
            - a[2] * y[n - 2]
        )
    return np.clip(y, -32768, 32767).astype(np.int16)


def ler_pcm(caminho_arquivo):
    """
    Lê um arquivo PCM e retorna os dados de áudio como numpy array.

    Parâmetros:
    - caminho_arquivo: Caminho do arquivo PCM.

    Retorna:
    - Array numpy com os dados do sinal.
    """
    with open(caminho_arquivo, "rb") as f:
        raw_data = f.read()
    return np.frombuffer(raw_data, dtype=np.int16)


def salvar_pcm(nome_arquivo, sinal):
    """
    Salva o sinal no formato PCM (16 bits).
    """
    with open(nome_arquivo, "wb") as f:
        f.write(sinal.tobytes())
    print(f'Sinal salvo como "{nome_arquivo}".')


def criar_pasta_resultados(pasta="Resultados"):
    """
    Cria uma pasta para salvar os resultados, se não existir.
    """
    if not os.path.exists(pasta):
        os.makedirs(pasta)
    return pasta


def calcular_atenuacao(w, h, frequencias):
    """
    Calcula a atenuação em dB para frequências especificadas.

    Parâmetros:
    - w: Frequências calculadas por freqz.
    - h: Resposta em frequência calculada por freqz.
    - frequencias: Lista de frequências de interesse.

    Retorna:
    - Dicionário com frequências como chave e atenuações em dB como valor.
    """
    attenuations = {}
    for freq in frequencias:
        idx = np.argmin(np.abs(w - freq))
        magnitude = np.abs(h[idx])
        attenuations[freq] = 20 * np.log10(magnitude) if magnitude > 0 else -np.inf
    return attenuations


# Configurações do filtro
fs = 8000  # Taxa de amostragem (8 kHz)
f0 = 400   # Frequência central do notch (Hz)
q = 30     # Fator de qualidade
frequencies = [200, 1000]  # Frequências de interesse para análise

# Calcular coeficientes
b, a = calculate_notch_coefficients(fs, f0, q)

# Criar pasta para resultados
pasta_resultados = criar_pasta_resultados("IIR_Rejeita_Faixa/Reultados")

# Carregar sinal de entrada
caminho_arquivo = "IIR_Rejeita_Faixa/Q2_voz_ruido.pcm"
input_signal = ler_pcm(caminho_arquivo)

# Aplicar filtro
output_signal = apply_filter(input_signal, b, a)

# Salvar sinal filtrado
salvar_pcm(os.path.join(pasta_resultados, "Sai_M2_voz_ruido.pcm"), output_signal)

# Gráfico da resposta em frequência com marcadores
w, h = freqz(b, a, worN=8000, fs=fs)
attenuations = calcular_atenuacao(w, h, frequencies)
plt.figure(figsize=(10, 6))
plt.plot(w, 20 * np.log10(np.abs(h)), label="Resposta em Frequência")
for freq, att in attenuations.items():
    idx = np.argmin(np.abs(w - freq))
    plt.annotate(f"{att:.2f} dB", xy=(w[idx], 20 * np.log10(np.abs(h[idx]))),
                 xytext=(w[idx] + 200, 20 * np.log10(np.abs(h[idx])) - 10),
                 arrowprops=dict(facecolor='red', arrowstyle="->"))
plt.title("Resposta em Frequência do Filtro")
plt.xlabel("Frequência (Hz)")
plt.ylabel("Magnitude (dB)")
plt.grid()
plt.legend()
plt.tight_layout()
plt.savefig(os.path.join(pasta_resultados, "resposta_frequencia.png"))
plt.close()

# Gráfico de entrada e saída
tempo = np.linspace(0, len(input_signal) / fs, len(input_signal))
plt.figure(figsize=(10, 6))
plt.subplot(2, 1, 1)
plt.plot(tempo, input_signal, label="Entrada", color="orange")
plt.title("Sinal de Entrada")
plt.grid()

plt.subplot(2, 1, 2)
plt.plot(tempo, output_signal, label="Saída", color="green")
plt.title("Sinal de Saída")
plt.grid()

plt.tight_layout()
plt.savefig(os.path.join(pasta_resultados, "sinais_entrada_saida.png"))
plt.close()

# Gráfico de polos e zeros
zeros = np.roots(b)
poles = np.roots(a)
plt.figure(figsize=(6, 6))
plt.scatter(np.real(zeros), np.imag(zeros), label="Zeros", color="blue")
plt.scatter(np.real(poles), np.imag(poles), label="Polos", color="red")
# Círculo unitário
unit_circle = np.exp(1j * np.linspace(0, 2 * np.pi, 100))
plt.plot(np.real(unit_circle), np.imag(unit_circle), linestyle="--", color="black", label="Círculo Unitário")
plt.axhline(0, color="black", linewidth=0.5)
plt.axvline(0, color="black", linewidth=0.5)
plt.title("Polos e Zeros do Filtro")
plt.xlabel("Parte Real")
plt.ylabel("Parte Imaginária")
plt.grid()
plt.legend()
plt.tight_layout()
plt.savefig(os.path.join(pasta_resultados, "polos_zeros.png"))
plt.close()

print(f"Gráficos e resultados salvos na pasta '{pasta_resultados}'.")

File: M2 Atividade/IIR_Rejeita_Faixa/rejeita_faixa.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define FS 8000  // Taxa de amostragem (8 kHz)
#define F0 400   // Frequência central (Hz)
#define Q 30     // Fator de qualidade
#define PI 3.14159265358979323846

// Estrutura para armazenar coeficientes do filtro
typedef struct {
    double b0, b1, b2;  // Numerador
    double a1, a2;      // Denominador
} FilterCoefficients;

// Função para calcular os coeficientes do filtro
FilterCoefficients calculate_notch_coefficients(double fs, double f0, double q) {
    FilterCoefficients coeffs;
    double omega = 2 * PI * f0 / fs;  // Frequência angular
    double alpha = sin(omega) / (2 * q);

    // Coeficientes do numerador
    coeffs.b0 = 1;
    coeffs.b1 = -2 * cos(omega);
    coeffs.b2 = 1;

    // Coeficientes do denominador
    coeffs.a1 = -2 * cos(omega);
    coeffs.a2 = 1 - alpha;

    // Normalizar todos os coeficientes pelo denominador a0 = 1 + alpha
    double a0 = 1 + alpha;
    coeffs.b0 /= a0;
    coeffs.b1 /= a0;
    coeffs.b2 /= a0;
    coeffs.a1 /= a0;
    coeffs.a2 /= a0;

    return coeffs;
}

void apply_filter(const short *input, short *output, int length, FilterCoefficients coeffs) {
    double x[3] = {0};  // Buffer de entrada
    double y[3] = {0};  // Buffer de saída

    for (int n = 0; n < length; n++) {
        // Atualizar buffers
        x[2] = x[1];
        x[1] = x[0];
        x[0] = input[n];

        y[2] = y[1];
        y[1] = y[0];

        // Aplicar equação do filtro
        y[0] = coeffs.b0 * x[0] + coeffs.b1 * x[1] + coeffs.b2 * x[2]
             - coeffs.a1 * y[1] - coeffs.a2 * y[2];

        // Converter para saída 16-bit
        output[n] = (short)y[0];
    }
}

int main() {
    // Calcular os coeficientes do filtro
    FilterCoefficients coeffs = calculate_notch_coefficients(FS, F0, Q);

    // Imprimir coeficientes calculados
    printf("Coeficientes do filtro:\n");
    printf("b0 = %.6f, b1 = %.6f, b2 = %.6f\n", coeffs.b0, coeffs.b1, coeffs.b2);
    printf("a1 = %.6f, a2 = %.6f\n", coeffs.a1, coeffs.a2);

    // Abrir arquivos de entrada e saída
    FILE *input_file = fopen("C:/Users/jorge/Desktop/Nova pasta/IIR_Rejeita_Faixa/Q2_voz_ruido.pcm", "rb");
    FILE *output_file = fopen("C:/Users/jorge/Desktop/Nova pasta/IIR_Rejeita_Faixa/Sai_C_M2_voz_ruido.pcm", "wb");

    if (!input_file || !output_file) {
        printf("Erro ao abrir arquivos.\n");
        return 1;
    }

    // Determinar tamanho do arquivo de entrada
    fseek(input_file, 0, SEEK_END);
    int length = ftell(input_file) / sizeof(short);  // Número de amostras
    fseek(input_file, 0, SEEK_SET);

    // Alocar memória para os sinais de entrada e saída
    short *input = malloc(length * sizeof(short));
    short *output = malloc(length * sizeof(short));

    if (input == NULL || output == NULL) {
        printf("Erro ao alocar memória.\n");
        fclose(input_file);
        fclose(output_file);
        return 1;
    }

    fread(input, sizeof(short), length, input_file);

    // Aplicar filtro
    apply_filter(input, output, length, coeffs);

    fwrite(output, sizeof(short), length, output_file);

    // Limpeza
    free(input);
    free(output);
    fclose(input_file);
    fclose(output_file);

    printf("Processamento concluído. Arquivo salvo como 'Sai_C_M2_voz_ruido.pcm'.\n");
    return 0;
}
